Add mention selectors to OpenCtx provider protocol

thenamankumar opened the pull request:
This PR modifies the `MetaResult` to the following:

- ~`mentions.kind` is the list of types of mention items the provider can return. For example github pr, issues, code search results.~
- `mentions.selectors` is the list of regex patterns, which if matches with the mention text would mean that the provider can return items for that. We will use it to get context items from all the supported providers for when the user pasts in a URL after @. 
- Whey we do not have `features.mentions: boolean | {}` ? because then you can not do `meta?.features?.mentions?.selector`. We will always have to check of the mentions type first. 
- Accordingly the `meta.selector` is moved to `meta.features.annotations.selectors`. 

``` typescript
export interface MetaResult {
    /**
     * The name of the provider.
     */
    name: string

        /**
         * Configuration for the mentions feature.
         */
        mentions?: {
            /**
             * A list of patterns matching the mention text for which the provider can return mentions
             */
            selectors?: MentionSelector[]
        }
        /**
         * Configuration for the annotations feature.
         */
        annotations?: {
            /**
             * A list of patterns matching the mention text for which the provider can return mentions
             */
            selectors?: AnnotationSelector[]
        }
}
```

sqs commented on lib/protocol/src/openctx-protocol.schema.json:
```diff
@@ -105,31 +114,58 @@
+      "type": "object",
+      "description": "A mention kind supported by the provider.",
+      "additionalProperties": false,
+      "required": ["id", "title"],
```
title should not be required (simplifies mock/test code)

thenamankumar replied:
Removed mention kind.

sqs commented on lib/protocol/src/openctx-protocol.schema.json:
```diff
@@ -105,31 +114,58 @@
       "tsType": "Record<string, never>"
     },
+    "MentionSelector": {
+      "description": "List of regex patterns matching the mention text for which the provider can return mentions.",
```
Describe more what this means, basically what you just told me: This is about supporting a 2nd way of the user selecting a mention. After typing @, they can select the mention kind OR they can type in a full URL (or something else I guess). The second way lets them bypass choosing the mention kind.

thenamankumar replied:
done

sqs commented on lib/protocol/src/openctx-protocol.schema.json:
```diff
@@ -140,8 +176,49 @@
+              "type": "object",
+              "additionalProperties": false,
+              "properties": {
+                "implements": {
```
Can we remove this? If there is an object here, it is implemented.

```
const implementsMentions = !!meta.features.mentions
```

Ie when would it ever be false?

thenamankumar replied:
I thought `!!{} = false`, but it isn't. I'll remove this.

thenamankumar replied:
done

sqs reviewed (approved):
but also check w/other people who have made providers recently to make sure both this makes sense and the docs make sense (non-blocking)

keegancsmith commented:
High level comments, just looked at PR description

- implements seems duplicative. Can't we just check if the mentions field is present?
- Does the feature nesting provide any value? Why not just have a high level mentions and annotations field on meta. This is just a taste perspective on "unnecessary nesting"
- The description of selector seems a bit vague. Additionally regexp isn't a universal thing, but there are many different flavours. Given we want HTTP support, maybe mention the regex is a javascript regex?
- Given we are creating behaviour changes based on the current message, it feels like `message` should be an optional part of `MentionParams`.
- Can't we just always call `mentions`, what is a concrete example where its useful to do this via a selector vs the mention function call itself deciding if it should provide mentions (assuming we now pass in message)

thenamankumar commented:
@sqs @keegancsmith Update:

- Removed the mentions.kinds altogether, @eseliger and I are exploring an idea of rather multiplexing at the provider level, which means just returning multiple providers for a single package for different things. It makes sense and we will discuss it separately. So this PR only implements mentions.selectors now. 
- @keegancsmith I have removed `meta.features` nesting.
- @sqs I have removed `implements`. 
- @keegancsmith @sqs I have updated the description for the mention selectors. Please provides ideas if wish to improve it further and we can always address is separately. 
- @keegancsmith we can not always call mentions on all the providers. This is being done to bypass provider selection. For example a user writes `@https://github.com/sourcegraph/cody/pull/123`, the Github & Repositories provider can say yes, we recognise it (based on a cheap client side regex match without calling provider as we have meta info beforehand), and then we can make `mentions` call to only those 2 providers and list the mentions, without the user having to select a provider at all. Also the same reason why we have selectors for annotations.

keegancsmith commented:
@thenamankumar do you have a PR yet for cody? This is a backwards incompatible change, so for example against current versions of cody my in development providers do not show up (since they no longer set features.mentions)
